src/complex_fluids/CFINSForcing.cpp

```diff
@@ -0,0 +1,952 @@
+                                                                 false,
+                                                                 d_conc_bc_coefs,
+                                                                 NULL,
+                                                                 d_interp_type);
+    HierarchyGhostCellInterpolation ghost_fill_op;
+    ghost_fill_op.initializeOperatorState(ghost_cell_components, hierarchy);
+    ghost_fill_op.fillData(data_time);
+
+    // Check to ensure conformation tensor is positive definite
+    d_positive_def = true;
+    checkPositiveDefinite(d_W_scratch_idx, d_W_cc_var, data_time, initial_time);
+    int temp = d_positive_def ? 1 : 0;
+    d_positive_def = SAMRAI_MPI::maxReduction(temp) == 1 ? true : false;
+    plog << "Conformation tensor is " << (d_positive_def ? "SPD" : "NOT SPD") << "\n";
+    if (d_error_on_spd && !d_positive_def)
+    {
+        TBOX_ERROR("ERROR: \n "
+                   << "    CONFORMATION TENSOR IS NOT SPD! \n\n");
+    }
+
+    // Check max and min determinant
+    if (d_log_det)
+    {
+        d_max_det = 0.0;
+        d_min_det = std::numeric_limits<double>::max();
+        findDeterminant(d_W_scratch_idx, d_W_cc_var, data_time, initial_time);
+        d_max_det = SAMRAI_MPI::maxReduction(d_max_det);
+        d_min_det = SAMRAI_MPI::minReduction(d_min_det);
+        plog << "Largest det:  " << d_max_det << "\n";
+        plog << "Smallest det: " << d_min_det << "\n";
+    }
+
+    // Compute Div W on each patch level
+    d_min_norm = std::numeric_limits<double>::max();
+    d_max_norm = 0.0;
+    for (int level_num = coarsest_ln; level_num <= finest_ln; ++level_num)
+    {
+        setDataOnPatchLevel(data_idx, var, hierarchy->getPatchLevel(level_num), data_time, initial_time);
+    }
+    // Output largest and smallest max norm of Div W
+    if (d_log_divW || d_divW_rel_tag)
+    {
+        SAMRAI_MPI::maxReduction(d_max_norm);
+        SAMRAI_MPI::minReduction(d_min_norm);
+        plog << "Largest max norm of Div W:  " << d_max_norm << "\n";
+        plog << "Smallest max norm of Div W: " << d_min_norm << "\n";
+    }
+
+    for (int level_num = coarsest_ln; level_num <= finest_ln; ++level_num)
+    {
+        Pointer<PatchLevel<NDIM> > level = hierarchy->getPatchLevel(level_num);
+        if (level->checkAllocated(d_W_scratch_idx)) level->deallocatePatchData(d_W_scratch_idx);
+    }
+    return;
+} // End setDataOnPatchHierarchy
+
+void
+CFINSForcing::setDataOnPatchLevel(const int data_idx,
+                                  Pointer<Variable<NDIM> > var,
+                                  Pointer<PatchLevel<NDIM> > level,
+                                  const double data_time,
+                                  const bool initial_time)
+{
+    if (initial_time)
+    {
+        // TODO: The integrators do not call setDataOnPatchHierarchy for the initial iteration. This is a workaround for
```

Comment:
@abarret can you create an issue outlining these issues?

Reply:
Yes.